chat.js

```diff
@@ -727,10 +729,16 @@ class CommandContext {
 				return false;
 			}
 
-			// If the corresponding config option is set, non-AC users cannot send links, except to staff.
-			if (Config.restrictLinks && !user.autoconfirmed && message.match(Chat.linkRegex)) {
-				if (!(targetUser && targetUser.can('lock'))) {
-					this.errorReply("Your account must be autoconfirmed to send links to other users, except for global staff.");
+			// If the corresponding Config option is set, non-AC users cannot send links, except to staff.
+			const links = message.match(Chat.linkRegex);
+			if (Config.restrictLinks && !user.autoconfirmed && links) {
+				const whitelisted = links.every(link => {
+					const [, subdomain, domain] = /^(?:http\:\/\/|https:\/\/)?((?:[^\.\/]*\.)?([^\.\/]*\.[^\.\/]*))\.?($|\/|:)/i.exec(link.toLowerCase()) || [];
+					if (!subdomain) return false;
+					return AC_LINK_WHITELIST.includes(subdomain) || AC_LINK_WHITELIST.includes(`*.${domain}`) || (subdomain.startsWith("www.") && AC_LINK_WHITELIST.includes(domain));
```

Comment:
Also, this code will let through `www.this-is-a-spam-server.psim.us`

Reply:
I think you guys are thinking too hard about "how do we let through links we want to let through" and not hard enough about "what links do we not want to let through can slip through these cracks?"

Reply:
```
>> /^(?:http\:\/\/|https:\/\/)?((?:[^\.\/]*\.)?([^\.\/]*\.[^\.\/]*))\.?($|\/|:)/i.exec("www.this-is-a-spam-server.psim.us".toLowerCase())
<< null
```

@Zarel the regex doesn't match on a domain of four groups. That's why I excluded dots, that forces the url to consist of either two or three groups.

Feel free to solve it differently if you don't trust this code.

Reply:
I mean, it's also wrong for it not to match on a domain of four groups.... Those do happen sometimes!